src/platforms/esp32/components/avm_builtins/storage_nif.c

```diff
@@ -0,0 +1,340 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR LGPL-2.1-or-later
+ */
+
+#include <sdkconfig.h>
+#ifdef CONFIG_AVM_ENABLE_STORAGE_NIFS
+
+#include <atom.h>
+#include <defaultatoms.h>
+#include <erl_nif.h>
+#include <erl_nif_priv.h>
+#include <interop.h>
+#include <memory.h>
+#include <nifs.h>
+#include <term.h>
+
+#include "esp32_sys.h"
+
+#include <stdlib.h>
+
+// file write test
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include <driver/sdmmc_host.h>
+#include <driver/sdspi_host.h>
+#include <esp_vfs_fat.h>
+
+#include <trace.h>
+
+#include "spi_driver.h"
+
+#define TAG "storage_nif"
+
+// TODO: allow ro option
+enum mount_type
+{
+    Unmounted,
+    FATSPIFlash,
+    FATSDSPI,
+    FATSDMMC
+};
+
+struct MountedFS
+{
+    char *base_path;
+    enum mount_type mount_type;
+    union
+    {
+        sdmmc_card_t *card;
+        wl_handle_t wl;
+    } handle;
+};
+
+static void mounted_fs_dtor(ErlNifEnv *caller_env, void *obj);
+
+const ErlNifResourceTypeInit mounted_fs_resource_type_init = {
+    .members = 1,
+    .dtor = mounted_fs_dtor
+};
+
+static term make_esp_error_tuple(esp_err_t err, Context *ctx)
+{
+    if (UNLIKELY(memory_ensure_free(ctx, TUPLE_SIZE(2)) != MEMORY_GC_OK)) {
+        RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+    }
+    term result = term_alloc_tuple(2, &ctx->heap);
+    term_put_tuple_element(result, 0, ERROR_ATOM);
+    term_put_tuple_element(result, 1, esp_err_to_term(ctx->global, err));
+    return result;
+}
+
+static void opts_to_fatfs_mount_config(term opts_term, esp_vfs_fat_mount_config_t *mount_config)
+{
+    mount_config->format_if_mount_failed = true;
+    mount_config->max_files = 8;
+    mount_config->allocation_unit_size = 512;
+    // TODO: make it configurable: disk_status_check_enable = false
+}
+
+static term nif_esp_mount(Context *ctx, int argc, term argv[])
+{
+    GlobalContext *glb = ctx->global;
+    struct ESP32PlatformData *platform = glb->platform_data;
+
+    term source_term = argv[0];
+    term target_term = argv[1];
+    term filesystem_type_term = argv[2];
+    term opts_term = argv[3];
+
+    int str_ok;
+    char *source = interop_term_to_string(source_term, &str_ok);
+    if (!str_ok) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    char *target = interop_term_to_string(target_term, &str_ok);
+    if (!str_ok || strlen(target) > 8) {
+        RAISE_ERROR(BADARG_ATOM);
```

Comment:
Shouldn't we free `source` here before returning an invalid term?

Reply:
Yes, and sorry for the pedantic repeated mistake, I think I fixed all of them.